### ISSUE TITLE
Additional information on `run_model` method from the `MetaICLModel` class

### ISSUE THREAD

**Issue by MicPie** [facebookresearch/MetaICL#9]
Hi @shmsw25 

thank you very much for publishing the MetaICL code base.

I'm currently going through the code to understand it better and I stumbled over the following lines in the `MetaICLModel` class of the `run_model` method:
https://github.com/facebookresearch/MetaICL/blob/main/metaicl/model.py#L273 (removal of the last element)
https://github.com/facebookresearch/MetaICL/blob/main/metaicl/model.py#L277:L278 (removal of the first element)

I'm not sure why the last and first element is removed, but my guess would be that it is due to the addition of newlines or spaces in the `_prepro_each_datapoint` function (https://github.com/facebookresearch/MetaICL/blob/main/metaicl/data.py#L116)?

(Another guess is that it is related to EOS and BOS tokens, but those are not used in the `MetaICLData` class because in the `prepro_sentence_pair_single` function they are commented out https://github.com/facebookresearch/MetaICL/blob/main/metaicl/data.py#L457:L460 )

Maybe I'm missing something important to understand the `run_model`method?

**Comment by shmsw25** [comment]
Thanks for the question. This is because the forward pass of `self.model` returns the logit where i-th element is the probability given w_1...w_i (w_i: i-th input; note that w_1...w_i includes both "input" and "output" of the task, which we distinguish through `label_mask`). So for calculating P(w_{i+1}|w_1...w_i), we need to shift `labels` to the left. So we discard the very first word of `label` (which won't matter anyways because the very first element of `label_mask` is always zero), and then discard the very last of `output.logit` since we don't care about what comes next after all the input and the output has been passed.

A running example that might help understand how it works: let's say the input is ["hello"] and the output is ["word"]. Then, 
`labels` is ["hello", "world"] and `label_mask` is [0, 1]. And `output.logit` is a 2-by-(vocab size) matrix that represents the next-token probability distribution: `[P(y|"hello"), P(y|"hello world")`. Therefore, by dropping the first of label and the last of `output.logit`, we calculate `P("world"|"hello")`.

Let me know if you have any further questions.

**Comment by MicPie** [comment]
Thank you for your detailed explanation. I was thinking in the wrong direction. :-)

PS: The simple code outlined here for the AR LM task made it then click also a 2. time: https://github.com/lucidrains/x-transformers/blob/main/x_transformers/autoregressive_wrapper.py#L113:L126
(they also the last token is removed before running it through the model)

**Comment by thesofakillers** [comment]
Question: How does this work for cases where the input and outputs consist of multiple tokens?

Consider the example where input is "hello there" and output is "beautiful world"?

Then 
- `output.logits` is `[P(y|"hello"), P(y|"hello there"), "P(y|"hello there beautiful"), "P(y|"hello there beautiful world")]`
- `labels` is `['hello', 'there', 'beautiful', 'world']`
- `label_mask` is `[0, 0, 1, 1]`

By dropping the last of output.logits and the first of labels, wouldn't you get

`p("there beautiful world" | "hello there beautiful")`?

Aren't we interested in 

`p("beatiful world" | "hello there")`?


Is this simply accounted for with the label mask? 


thanks! sorry for the many questions on this repo from me

**Comment by ke-01** [comment]
> Thanks for the question. This is because the forward pass of `self.model` returns the logit where i-th element is the probability given w_1...w_i (w_i: i-th input; note that w_1...w_i includes both "input" and "output" of the task, which we distinguish through `label_mask`). So for calculating P(w_{i+1}|w_1...w_i), we need to shift `labels` to the left. So we discard the very first word of `label` (which won't matter anyways because the very first element of `label_mask` is always zero), and then discard the very last of `output.logit` since we don't care about what comes next after all the input and the output has been passed.
> 
> A running example that might help understand how it works: let's say the input is ["hello"] and the output is ["word"]. Then, `labels` is ["hello", "world"] and `label_mask` is [0, 1]. And `output.logit` is a 2-by-(vocab size) matrix that represents the next-token probability distribution: `[P(y|"hello"), P(y|"hello world")`. Therefore, by dropping the first of label and the last of `output.logit`, we calculate `P("world"|"hello")`.
> 
> Let me know if you have any further questions.

Thank you very much for your answer, but I want to ask one more question. 
By dropping the first of label and the last of output.logit, can it be understood that we use the label position of input_ids and the previous position of the label in the logits generated by the model to obtain the loss. Then we select the label corresponding to the ids that is more likely to be generated by the previous position of the label in the logits generated by the model as the prediction result?
in the following cases：(labels == input_ids)
https://github.com/facebookresearch/MetaICL/blob/ec0f1c199965f54d4b4d35556e5ca9ba0a6719ab/metaicl/model.py#L278

**Comment by shmsw25** [comment]
@thesofakillers (sorry for the late reply. I assume you already figured it out, but just to respond) and @ke-01, I believe both of you are asking the same question, so I'll respond altogether. Let me use the running example @thesofakillers provided -- Let's say the input is `hello there` and the output is `beautiful world`. 

* output.logits is `[P(y|"hello"), P(y|"hello there"), P(y|"hello there beautiful"), P(y|"hello there beautiful world")]`
* labels is `['hello', 'there', 'beautiful', 'world']`
* label_mask is `[0, 0, 1, 1]`

We drop the last of outputs logits, and drop the first of the labels & label_mask.
* output.logits now is `[P(y|"hello"), P(y|"hello there"), P(y|"hello there beautiful")`
* labels now is `['there', 'beautiful', 'world']`
* label_mask now is `[0, 1, 1]`

Thus, this computes the following.
* `P(y|"hello")` with respect to `there` --> but this is eventually ignored because the corresponding label mask is 0.
* `P(y|"hello there")` with respect to `beautiful`
* `P(y|"hello there beautiful")` with respect to `world`

Therefore, it ends up with `P(beautiful world | hello there)`.  Please refer to [this line](https://github.com/facebookresearch/MetaICL/blob/main/metaicl/model.py#L285) to see how exactly `label_mask` is applied to the loss.

Hope this helps!
